File: ros2_ws/src/teleop_bridge/teleop_bridge/utils/logger.py
import json
import time
import os

class Logger:
    """
    [Bridge Integrated Pipeline Profiling]
    -----------------------------------------------------------------------------
    < Python Domain >
    T0_a (capture_done) : Image arrival from ROS2 node.
    T0_b (send_request) : WebRTC track retrieval (embedded in barcode).

    < Unity Domain (Received via DataChannel/API) >
    T1 (recv_request)    : Unity detected the frame and requested GPU readback.
    T2 (recv_arrival)    : GPU-to-CPU transfer complete.
    T3 (recv_decode_done): Barcode scan complete (Frame ID verified).
    T4 (recv_sync_done)  : Multi-stream synchronization finalized in Unity.

    < Analysis Formulae >
    - Internal Bridge Delay : T0_b - T0_a
    - Network + Encode Delay: T1 - T0_b (Includes Clock Offset)
    - Unity GPU Overhead   : T2 - T1
    - Final E2E Latency    : T4 - T0_a (Corrected by Offset)
    -----------------------------------------------------------------------------
    """
    
    def __init__(self, filename="log.jsonl"):
        self.session_id = int(time.time())
        self.file = open(filename, "w")

    def log(self, stage, frame_id, timestamp=None, extra=None):
        if timestamp is None:
            timestamp = time.time()

        entry = {
            "session": self.session_id,
            "stage": stage,
            "frame_id": frame_id,
            "ts": timestamp
        }

        if extra:
            entry.update(extra)
        
        self.file.write(json.dumps(entry) + "\n")
        self.file.flush()

logger = Logger("ros_log.jsonl")

File: unity/headset_layer/Assets/Scripts/Utils/Logger.cs
using System;
using System.IO;
using UnityEngine;

namespace Utils
{
    /*
     * [Unity Pipeline Profiling Definitions]
     * -----------------------------------------------------------------------------
     * T0 (Python Side) : Birth time of the frame (embedded in barcode).
     * * T1 (recv_request)    : Requesting pixel data from GPU texture (AsyncGPUReadback).
     * Identifies the exact moment Unity's rendering engine
     * received the WebRTC frame.
     * * T2 (recv_arrival)    : GPU-to-CPU memory transfer (DMA) completed.
     * Measures the hardware/driver overhead of data moving to CPU.
     * * T3 (recv_decode_done): Barcode scanning & FrameID/T0 recovery finished.
     * Measures the computational cost of the decoding logic.
     * * T4 (recv_sync_done)  : All required streams (RGB/Depth) matched at the Sync Barrier.
     * Identifies "waiting time" caused by network jitter or
     * unbalanced stream delivery.
     * -----------------------------------------------------------------------------
     */
    public class Logger
    {
        private StreamWriter _writer;
        private long _sessionId;
    
        public  Logger(string fileName = "unity_log.json1")
        {
            _sessionId = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            _writer = new StreamWriter(fileName, false);
        }

        public void Log(string stage, long frameId, long unityTime, string stream, long bridge_ts)
        {
            var json = JsonUtility.ToJson(new LogEntry
            {
                session = _sessionId,
                stage = stage,
                frame_id = frameId,
                unityTime = unityTime,
                stream = stream,
                bridgeTime = bridge_ts
            });
        
            _writer.WriteLine(json);
            _writer.Flush();
        }
    }

    public class LogEntry
    {
        public long session;
        public string stage;
        public long frame_id;
        
        /// <summary>
        /// Unity Timestamp
        /// </summary>
        public long unityTime; 
        public string stream;

        /// <summary>
        /// Bridge Timestamp 
        /// </summary>
        public long bridgeTime;
    }
}